modules/darwinosl/darwinosl-source.m

```diff
@@ -0,0 +1,559 @@
+            If there are multiple occurrences of the same time, the method returns the earliest occurrence.
+            So, in a worst case there might be log duplications on restart from bookmark
+  */
+  self->log_source_position.log_position = [nextLogEntry.date timeIntervalSince1970];
+  self->log_source_position.last_msg_hash = g_str_hash(log_string);
+
+  LogSource *worker = (LogSource *) self->super.worker;
+  Bookmark *bookmark = ack_tracker_request_bookmark(worker->ack_tracker);
+  darwinosl_source_persist_fill_bookmark(self->log_source_persist, bookmark, self->log_source_position);
+  msg_trace("darwinosl: Bookmark created",
+            evt_tag_printf("position", "%.6f", self->log_source_position.log_position),
+            evt_tag_printf("last message hash", "%u", self->log_source_position.last_msg_hash),
+            evt_tag_printf("filter predicate hash", "%u", self->log_source_position.last_used_filter_predicate_hash));
+
+  LogMessage *msg;
+  gsize msg_len = _log_message_from_string(log_string, self->options.format_options, &msg);
+  LogThreadedFetchResult result = {THREADED_FETCH_SUCCESS, msg};
+  _log_reader_insert_msg_length_stats(self, msg_len);
+  return result;
+}
+
+static void
+_send(LogThreadedSourceDriver *s, LogMessage *msg)
+{
+  log_threaded_source_blocking_post(s, msg);
+}
+
+static void
+_request_exit(LogThreadedSourceDriver *s)
+{
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *) s;
+  g_atomic_counter_set(&self->exit_requested, TRUE);
+}
+
+/* TODO: Using some iv_timer or similar instead ?! */
+static void
+_sleep(DarwinOSLogSourceDriver *self, gdouble wait_time)
+{
+  const useconds_t microseconds_in_second = 1000 * 1000;
+  const useconds_t min_sleep_time = 1;
+  const useconds_t def_sleep_time = 1000;
+  const useconds_t sleep_time = MAX(min_sleep_time, (useconds_t) MIN(def_sleep_time, wait_time * microseconds_in_second));
+  GTimeVal now, last_check;
+
+  g_get_current_time(&now);
+  last_check = now;
+
+  while (false == g_atomic_counter_get(&self->exit_requested))
+    {
+      usleep(sleep_time);
+
+      gdouble diff = g_time_val_diff(&last_check, &now) / (gdouble) microseconds_in_second;
+      if (diff >= wait_time)
+        break;
+      g_get_current_time(&last_check);
+    }
+}
+
+/* runs in a dedicated thread */
+static void
+_run(LogThreadedSourceDriver *s)
+{
+  DarwinOSLogSourceDriver *self = (DarwinOSLogSourceDriver *) s;
+
+  while (false == g_atomic_counter_get(&self->exit_requested))
+    {
+      @autoreleasepool
+      {
+        if (FALSE == _open_osl(s))
+          break;
+
+        while (false == g_atomic_counter_get(&self->exit_requested))
+          {
+            @autoreleasepool
+            {
+              LogThreadedFetchResult res = _fetch(s);
+              if (res.result == THREADED_FETCH_SUCCESS)
+                _send(s, res.msg);
+              else
+                break;
+            }
+            _sleep(self, 1.0 / self->options.fetch_delay);
```

Comment:
could we avoid floating point calculations here? We tried to avoid those in fast paths.
Maybe by precalculate it or used fixed mode arithmetic